includes/derivatives.inc

```diff
@@ -0,0 +1,131 @@
+<?php
+
+/**
+ * @file
+ * Holds derivative functionality for compound solution pack.
+ */
+
+/**
+ * Derivative creation function for compound parent object TN.
+ *
+ * Wraps the TN update/creation function and
+ * adds expected logging.
+ *
+ * @param AbstractObject $object
+ *   The compound object.
+ * @param bool $force
+ *   Boolean for forcing this derivative creation.
+ *
+ * @return array
+ *   Success message.
+ */
+function islandora_compound_object_create_thumbnail(AbstractObject $object, $force = TRUE) {
+  if ($force || !isset($object['TN'])) {
+    $parts = islandora_compound_object_get_parts($object->id);
+    if (!empty($parts)) {
+      $child = islandora_object_load($parts[0]);
+      islandora_compound_object_update_parent_thumbnail($object, $child);
+    }
+    if (isset($child) && $object['TN']->size == $child['TN']->size) {
```

Comment:
Adam is right
Something like:

```
     try {
        islandora_compound_object_update_parent_thumbnail($parent, $child);
      }
      catch (Exception $e) {
        return array(
          'success' => FALSE,
          'messages' => array(
            array(
              'message' => t('Failed to update @dsid on @parent.'),
              'message_sub' => array(
                '@dsid' => 'TN',
                '@parent' => $parent->id,
              ),
              'type' => 'watchdog',
              'severity' => WATCHDOG_ERROR,
            ),
          ),
        );
      }

      return array(
        'success' => TRUE,
        'messages' => array(
          array(
            'message' => t('Updated @dsid on @parent.'),
            'message_sub' => array(
              '@dsid' => 'TN',
              '@parent' => $parent->id,
            ),
            'type' => 'dsm',
            'severity' => 'status',
          ),
        ),
      );
    }
  }
```
And wrap islandora_compound_object_update_parent_thumbnail's logic in try catch?